lib/Serialization/ASTWriter.cpp

```diff
@@ -286,22 +286,28 @@ void ASTTypeWriter::VisitFunctionProtoType(const FunctionProtoType *T) {
 
   Record.push_back(T->isVariadic());
   Record.push_back(T->hasTrailingReturn());
+  Record.push_back(T->hasParamBounds());
   Record.push_back(T->getTypeQuals());
   Record.push_back(static_cast<unsigned>(T->getRefQualifier()));
   addExceptionSpec(T, Record);
+  Record.AddStmt(const_cast<BoundsExpr *>(T->getReturnBounds()));
 
   Record.push_back(T->getNumParams());
   for (unsigned I = 0, N = T->getNumParams(); I != N; ++I)
     Record.AddTypeRef(T->getParamType(I));
 
+  if (T->hasParamBounds())
+    for (unsigned I = 0, N = T->getNumParams(); I != N; ++I)
```

Comment:
I followed the coding style used in the prior loop over params.   I think this is an example of an overzealous compiler writer following a coding style that _might_ save a cycle or two.   I agree it is odd, but I'll keep consistency with the other loop.